### ISSUE TITLE
tokio-epoll-uring: "Cannot allocate memory" in staging on 2024-02-07

### ISSUE THREAD

**Issue by problame** [neondatabase/neon#6667]
* [Slack thread](https://neondb.slack.com/archives/C033RQ5SPDH/p1707314951137259)
* [Sentry link](https://neondatabase.sentry.io/issues/4952872723/?referrer=slack&notification_uuid=0eafcd10-02a1-406b-a314-f9b47f6f5b67&alert_rule_id=13489024&alert_type=issue)

Sentry-captured stack trace

<details>

```
OS Version: Linux 5.10.0-18-cloud-amd64 (None)
Report Version: 104


Application Specific Information:
called `Result::unwrap()` on an `Err` value: IoUringBuild(Os { code: 12, kind: OutOfMemory, message: "Cannot allocate memory" })

Thread 0 Crashed:
0   std                             0x55b2df65de36      std::sys_common::backtrace::__rust_end_short_backtrace (backtrace.rs:170)
1   <unknown>                       0x55b2df65f3a2      rust_begin_unwind (panicking.rs:645)
2   core                            0x55b2df68e7c5      core::panicking::panic_fmt (panicking.rs:72)
3   core                            0x55b2df68ed03      core::result::unwrap_failed (result.rs:1653)
4   core                            0x55b2de490643      [inlined] core::result::Result<T>::unwrap (result.rs:1077)
5   tokio_epoll_uring               0x55b2de490643      [inlined] tokio_epoll_uring::system::lifecycle::thread_local::thread_local_system::{{closure}}::{{closure}}::{{closure}} (thread_local.rs:16)
6   tokio                           0x55b2de490643      [inlined] tokio::sync::once_cell::OnceCell<T>::get_or_init::{{closure}} (once_cell.rs:368)
7   tokio_epoll_uring               0x55b2de490643      tokio_epoll_uring::system::lifecycle::thread_local::thread_local_system::{{closure}} (thread_local.rs:17)
8   pageserver                      0x55b2de4dc5e8      pageserver::virtual_file::open_options::OpenOptions::open::{{closure}} (open_options.rs:100)
9   pageserver                      0x55b2de4dd254      pageserver::virtual_file::VirtualFile::open_with_options::{{closure}} (virtual_file.rs:375)
10  pageserver                      0x55b2de4c96a3      [inlined] pageserver::virtual_file::VirtualFile::create::{{closure}} (virtual_file.rs:337)
11  pageserver                      0x55b2de4c96a3      [inlined] pageserver::tenant::storage_layer::delta_layer::DeltaLayerWriterInner::new::{{closure}} (delta_layer.rs:392)
12  pageserver                      0x55b2de4c96a3      [inlined] pageserver::tenant::storage_layer::delta_layer::DeltaLayerWriter::new::{{closure}} (delta_layer.rs:572)
13  pageserver                      0x55b2de4c96a3      pageserver::tenant::storage_layer::inmemory_layer::InMemoryLayer::write_to_disk::{{closure}} (inmemory_layer.rs:363)
14  tokio                           0x55b2de4b565c      [inlined] tokio::runtime::park::CachedParkThread::block_on::{{closure}} (park.rs:282)
15  tokio                           0x55b2de4b565c      [inlined] tokio::runtime::coop::with_budget (coop.rs:107)
16  tokio                           0x55b2de4b565c      [inlined] tokio::runtime::coop::budget (coop.rs:73)
17  tokio                           0x55b2de4b565c      [inlined] tokio::runtime::park::CachedParkThread::block_on (park.rs:282)
18  tokio                           0x55b2de4b565c      tokio::runtime::context::blocking::BlockingRegionGuard::block_on (blocking.rs:66)
19  tokio                           0x55b2de4054c0      [inlined] tokio::runtime::handle::Handle::block_on::{{closure}} (handle.rs:310)
20  tokio                           0x55b2de4054c0      tokio::runtime::context::runtime::enter_runtime (runtime.rs:65)
21  tokio                           0x55b2de48bebc      [inlined] tokio::runtime::handle::Handle::block_on (handle.rs:309)
22  pageserver                      0x55b2de48bebc      [inlined] pageserver::tenant::timeline::Timeline::create_delta_layer::{{closure}}::{{closure}} (timeline.rs:3086)
23  tokio                           0x55b2de48bebc      tokio::runtime::blocking::task::BlockingTask<T>::poll (task.rs:42)
24  tokio                           0x55b2de7c29b6      [inlined] tokio::runtime::task::core::Core<T>::poll::{{closure}} (core.rs:328)
25  tokio                           0x55b2de7c29b6      [inlined] tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut (unsafe_cell.rs:16)
26  tokio                           0x55b2de7c29b6      tokio::runtime::task::core::Core<T>::poll (core.rs:317)
27  tokio                           0x55b2de3481fc      [inlined] tokio::runtime::task::harness::poll_future::{{closure}} (harness.rs:485)
28  core                            0x55b2de3481fc      [inlined] core::panic::unwind_safe::AssertUnwindSafe<T>::call_once (unwind_safe.rs:272)
29  std                             0x55b2de3481fc      [inlined] std::panicking::try::do_call (panicking.rs:552)
30  std                             0x55b2de3481fc      [inlined] std::panicking::try (panicking.rs:516)
31  std                             0x55b2de3481fc      [inlined] std::panic::catch_unwind (panic.rs:142)
32  tokio                           0x55b2de3481fc      [inlined] tokio::runtime::task::harness::poll_future (harness.rs:473)
33  tokio                           0x55b2de3481fc      [inlined] tokio::runtime::task::harness::Harness<T>::poll_inner (harness.rs:208)
34  tokio                           0x55b2de3481fc      tokio::runtime::task::harness::Harness<T>::poll (harness.rs:153)
35  tokio                           0x55b2df541fdb      tokio::runtime::blocking::pool::Inner::run
36  std                             0x55b2df53b597      std::sys_common::backtrace::__rust_begin_short_backtrace
37  core                            0x55b2df53bdb9      core::ops::function::FnOnce::call_once{{vtable.shim}}
38  alloc                           0x55b2df6664f5      [inlined] alloc::boxed::Box<T>::call_once (boxed.rs:2007)
39  alloc                           0x55b2df6664f5      [inlined] alloc::boxed::Box<T>::call_once (boxed.rs:2007)
40  std                             0x55b2df6664f5      std::sys::unix::thread::Thread::new::thread_start (thread.rs:108)
41  <unknown>                       0x7f0bc3c48ea7      start_thread
42  <unknown>                       0x7f0bc3a1ca2f      clone
43  <unknown>                       0x0                 <unknown>
```

</details>

tl;dr: a `Handle::block_on` call inside `spawn_blocking`

Kernel version is `5.10.0-18-cloud-amd64`

=> As per our findings in https://github.com/neondatabase/neon/issues/6373#issuecomment-1905814391 , this means the process ran out of memlock rusage quota.

---

```[tasklist]
### Action Items
- [x] metric for memlocked usage; while at it, metrics for number of spawns?
- [x] build dashboard
- [x] investigate whether thread-local freeing works correctly with the block_on scenario
- [x] let peter run some more benchmarks & watch these metrics
- [ ] raise memlocked limit yet again to prevent this from re-occuring?
- [ ] discuss with team: given this can panic, should we rather abort() the process in that case?
- [ ] didn't happen again, not raisiong it further, changes on the write will reduce number of `spawn_blocking(Handle::block_on)` and hence system creations
```

```[tasklist]
### Impl
- [ ] https://github.com/neondatabase/neon/pull/6669
- [ ] https://github.com/neondatabase/tokio-epoll-uring/pull/44
- [ ] https://github.com/neondatabase/tokio-epoll-uring/pull/45
- [ ] https://github.com/neondatabase/neon/pull/6672
- [x] dashboard: https://neonprod.grafana.net/d/bdc5nq2nte680f/2024-02-tokio-epoll-uring-rollout
- [ ] https://github.com/neondatabase/neon/pull/6731
```
